# Машинное обучение. Лабораторная работа 1

Выполнил: _Коростелев Дмитрий Васильевич_

Группа:   _М8О-308Б-18_

## Настройка среды

In [1]:
%pip install numpy
%pip install pandas
%pip install sklearn

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


## Задача и датасет

Датасет взят с сайта : http://archive.ics.uci.edu/ml/datasets/Mushroom

В базе данных представлена информацию о видах грибов, их свойствах и особенностях. По описанию гриба при помощи искусственного интеллекта будем решать задачу классификации грибов на ядовитые и съедобные

## Датасет

In [2]:
import pandas as pds

df = pds.read_csv('agaricus-lepiota.data')

In [3]:
df

edible cap-shape cap-surface cap-color bruises odor gill-attachment  \
0         p         x           s         n       t    p               f   
1         e         x           s         y       t    a               f   
2         e         b           s         w       t    l               f   
3         p         x           y         w       t    p               f   
4         e         x           s         g       f    n               f   
5         e         x           y         y       t    a               f   
6         e         b           s         w       t    a               f   
7         e         b           y         w       t    l               f   
8         p         x           y         w       t    p               f   
9         e         b           s         y       t    a               f   
10        e         x           y         y       t    l               f   
11        e         x           y         y       t    a               f   
12        e         b           s         y       t    a               f   
13        p         x           y         w       t    p               f   
14        e         x           f         n       f    n               f   
15        e         s           f         g       f    n               f   
16        e         f           f         w       f    n               f   
17        p         x           s         n       t    p               f   
18        p         x           y         w       t    p               f   
19        p         x           s         n       t    p               f   
20        e         b           s         y       t    a               f   
21        p         x           y         n       t    p               f   
22        e         b           y         y       t    l               f   
23        e         b           y         w       t    a               f   
24        e         b           s         w       t    l               f   
25        p         f           s         w       t    p               f   
26        e         x           y         y       t    a               f   
27        e         x           y         w       t    l               f   
28        e         f           f         n       f    n               f   
29        e         x           s         y       t    a               f   
...     ...       ...         ...       ...     ...  ...             ...   
8094      e         b           s         g       f    n               f   
8095      p         x           y         c       f    m               f   
8096      e         k           f         w       f    n               f   
8097      p         k           y         n       f    s               f   
8098      p         k           s         e       f    y               f   
8099      e         k           f         w       f    n               f   
8100      e         f           s         n       f    n               a   
8101      p         k           s         e       f    s               f   
8102      e         x           s         n       f    n               a   
8103      e         k           s         n       f    n               a   
8104      e         k           s         n       f    n               a   
8105      e         k           s         n       f    n               a   
8106      e         k           s         n       f    n               a   
8107      e         x           s         n       f    n               a   
8108      p         k           y         e       f    y               f   
8109      e         b           s         w       f    n               f   
8110      e         x           s         n       f    n               a   
8111      e         k           s         w       f    n               f   
8112      e         k           s         n       f    n               a   
8113      p         k           y         e       f    y               f   
8114      p         f           y         c       f  

## Описание атрибутов и возможных значений: 

0. edible : p=poisonous, e=edible
1. cap-shape: bell=b,conical=c,convex=x,flat=f, knobbed=k,sunken=s
2. cap-surface: fibrous=f,grooves=g,scaly=y,smooth=s
3. cap-color: brown=n,buff=b,cinnamon=c,gray=g,green=r, pink=p,purple=u,red=e,white=w,yellow=y
4. bruises?: bruises=t,no=f
5. odor: almond=a,anise=l,creosote=c,fishy=y,foul=f, musty=m,none=n,pungent=p,spicy=s
6. gill-attachment: attached=a,descending=d,free=f,notched=n
7. gill-spacing: close=c,crowded=w,distant=d
8. gill-size: broad=b,narrow=n
9. gill-color: black=k,brown=n,buff=b,chocolate=h,gray=g, green=r,orange=o,pink=p,purple=u,red=e, white=w,yellow=y
10. stalk-shape: enlarging=e,tapering=t
11. stalk-root: bulbous=b,club=c,cup=u,equal=e, rhizomorphs=z,rooted=r,missing=?
12. stalk-surface-above-ring: fibrous=f,scaly=y,silky=k,smooth=s
13. stalk-surface-below-ring: fibrous=f,scaly=y,silky=k,smooth=s
14. stalk-color-above-ring: brown=n,buff=b,cinnamon=c,gray=g,orange=o, pink=p,red=e,white=w,yellow=y
15. stalk-color-below-ring: brown=n,buff=b,cinnamon=c,gray=g,orange=o, pink=p,red=e,white=w,yellow=y
16. veil-type: partial=p,universal=u
17. veil-color: brown=n,orange=o,white=w,yellow=y
18. ring-number: none=n,one=o,two=t
19. ring-type: cobwebby=c,evanescent=e,flaring=f,large=l, none=n,pendant=p,sheathing=s,zone=z
20. spore-print-color: black=k,brown=n,buff=b,chocolate=h,green=r, orange=o,purple=u,white=w,yellow=y
21. population: abundant=a,clustered=c,numerous=n, scattered=s,several=v,solitary=y
22. habitat: grasses=g,leaves=l,meadows=m,paths=p, urban=u,waste=w,woods=d

Разделим датасет на обучающую и тестовую выборки. Так как записей в датасете - 8123, разделим датасет в соотношении 6:1 

# Алгоритм KNN

## Подготовка датасета

Реализуем алгоритм KNN, сначала сопоставим всем качественным признакам числовые значения и будем работать с числовыми значениями.

In [4]:
edible_dict = { 'p' : 0, 'e' : 1 }
cap_shape_dict = { 'b' : 0, 'c' : 1, 'x' : 2, 'f' : 3, 'k' : 4, 's' : 5 }
cap_surface_dict = { 'f' : 0, 'g' : 1, 'y' : 2, 's' : 3 }
cap_color_dict = { 'n' : 0, 'b' : 1, 'c' : 2, 'g' : 3, 'r' : 4, 'p' : 5, 'u' : 6, 'e' : 7, 'w' : 8, 'y' : 9 }
bruises_dict = { 't' : 0, 'f' : 1 }
odor_dict = { 'a' : 0, 'l' : 1, 'c' : 2, 'y' : 3, 'f' : 4, 'm' : 5, 'n' : 6, 'p' : 7, 's' : 8 }
gill_attachment_dict = { 'a' : 0, 'd' : 1, 'f' : 2, 'n' : 3 }
gill_spacing_dict = { 'c' : 0, 'w' : 1, 'd' : 2 }
gill_size_dict = { 'b' : 0, 'n' : 1 }
gill_color_dict = { 'k' : 0, 'n' : 1, 'b' : 2, 'h' : 3, 'g' : 4, 'r' : 5, 'o' : 6, 'p' : 7, 'u' : 8, 'e' : 9, 'w' : 10, 'y' : 11 }
stalk_shape_dict = { 'e' : 0, 't' : 1 }
stalk_root_dict = { 'b' : 0, 'c' : 1, 'u' : 2, 'e' : 3, 'z' : 4, 'r' : 5, '?' : 6 }
stalk_surface_above_ring_dict = { 'f' : 0, 'y' : 1, 'k' : 2, 's' : 3 }
stalk_surface_below_ring_dict = { 'f' : 0, 'y' : 1, 'k' : 2, 's' : 3 }
stalk_color_above_ring_dict = { 'n' : 0, 'b' : 1, 'c' : 2, 'g' : 3, 'o' : 4, 'p' : 5, 'e' : 6, 'w' : 7, 'y' : 8 }
stalk_color_below_ring_dict = { 'n' : 0, 'b' : 1, 'c' : 2, 'g' : 3, 'o' : 4, 'p' : 5, 'e' : 6, 'w' : 7, 'y' : 8 }
veil_type_dict = { 'p' : 0, 'u' : 1 }
veil_color_dict = { 'n' : 0, 'o' : 1, 'w' : 2, 'y' : 3 }
ring_number_dict = { 'n' : 0, 'o' : 1, 't' : 2 }
ring_type_dict = { 'c' : 0, 'e' : 1, 'f' : 2, 'l' : 3, 'n' : 4, 'p' : 5, 's' : 6, 'z' : 7 }
spore_print_color_dict = { 'k' : 0, 'n' : 1, 'b' : 2, 'h' : 3, 'r' : 4, 'o' : 5, 'u' : 6, 'w' : 7, 'y' : 8 }
population_dict = { 'a' : 0, 'c' : 1, 'n' : 2, 's' : 3, 'v' : 4, 'y' : 5 }
habitat_dict = { 'g' : 0, 'l' : 1, 'm' : 2, 'p' : 3, 'u' : 4, 'w' : 5, 'd' : 6 }

def replace_attribute_on_number(row):
    row['edible'] = edible_dict[row['edible']] # 1
    row['cap-shape'] = cap_shape_dict[row['cap-shape']] # 2
    row['cap-surface'] = cap_surface_dict[row['cap-surface']] # 3
    row['cap-color'] = cap_color_dict[row['cap-color']] # 4
    row['bruises'] = bruises_dict[row['bruises']] # 5
    row['odor'] = odor_dict[row['odor']] # 6
    row['gill-attachment'] = gill_attachment_dict[row['gill-attachment']] # 7
    row['gill-spacing'] = gill_spacing_dict[row['gill-spacing']] # 8
    row['gill-size'] = gill_size_dict[row['gill-size']] # 9
    row['gill-color'] = gill_color_dict[row['gill-color']] # 10
    row['stalk-shape'] = stalk_shape_dict[row['stalk-shape']] # 11
    row['stalk-root'] = stalk_root_dict[row['stalk-root']] # 12
    row['stalk-surface-above-ring'] = stalk_surface_above_ring_dict[row['stalk-surface-above-ring']] # 13
    row['stalk-surface-below-ring'] = stalk_surface_below_ring_dict[row['stalk-surface-below-ring']] # 14
    row['stalk-color-above-ring'] = stalk_color_above_ring_dict[row['stalk-color-above-ring']] # 15
    row['stalk-color-below-ring'] = stalk_color_below_ring_dict[row['stalk-color-below-ring']] # 16
    row['veil-type'] = veil_type_dict[row['veil-type']] # 17
    row['veil-color'] = veil_color_dict[row['veil-color']] # 18
    row['ring-number'] = ring_number_dict[row['ring-number']] # 19
    row['ring-type'] = ring_type_dict[row['ring-type']] # 20
    row['spore-print-color'] = spore_print_color_dict[row['spore-print-color']] # 21
    row['population'] = population_dict[row['population']] # 22
    row['habitat'] = habitat_dict[row['habitat']] # 23


def change_atr_to_numbers_in_df(df):
    df_copy = df.copy()
    for index, row in df_copy.iterrows():
        replace_attribute_on_number(row)
    return df_copy

In [5]:
df_numeric = change_atr_to_numbers_in_df(df)
df_numeric = df_numeric.sample(frac=1)
sep = len(df_numeric)//7
df_testing = df_numeric.copy().iloc[0:sep]
df_learning = df_numeric.copy().iloc[sep:]
print(sep)

1160


In [6]:
df_numeric

edible cap-shape cap-surface cap-color bruises odor gill-attachment  \
7576      1         4           0         3       1    6               2   
7943      0         4           2         0       1    4               2   
6589      0         3           3         0       1    4               2   
1900      1         3           0         3       0    6               2   
2389      1         2           2         3       0    6               2   
2951      1         2           2         0       0    6               2   
2971      1         2           0         3       0    6               2   
4803      0         3           2         3       1    4               2   
7718      0         3           3         0       1    8               2   
5473      0         2           2         7       1    3               2   
462       1         0           3         8       0    0               2   
192       1         0           3         8       0    1               2   
596       1         0           3         9       0    1               2   
3759      1         3           0         7       0    6               2   
7844      1         4           3         0       1    6               0   
154       1         2           3         8       0    0               2   
6339      0         2           3         7       1    4               2   
2403      1         3           2         7       0    6               2   
6106      0         2           2         7       1    8               2   
4794      0         3           2         3       1    4               2   
1305      1         3           3         3       1    6               2   
523       0         2           3         0       0    7               2   
6905      1         0           3         0       1    6               0   
464       1         2           0         0       1    6               2   
2484      1         2           0         7       0    6               2   
1339      1         3           0         8       1    6               2   
5076      0         3           0         3       1    4               2   
5683      1         2           3         0       0    6               2   
5248      0         0           2         8       0    6               2   
6148      0         2           3         0       1    8               2   
...     ...       ...         ...       ...     ...  ...             ...   
3698      1         3           2         3       0    6               2   
7972      1         0           3         8       1    6               2   
3665      1         3           0         7       0    6               2   
1410      1         2           0         0       1    6               2   
7907      1         0           0         8       1    6               2   
2407      1         2           0         7       0    6               2   
1941      1         3           0         0       0    6               2   
2890      1         2           0         0       1    6               2   
2151      1         3           0         0       0    6               2   
1732      1         3           3         3       1    6               2   
6409      0         2           2         7       1    4               2   
6390      0         4           2         0       1    3               2   
8028      0         4           3         0       1    4               2   
4836      0         2           0         9       1    4               2   
6605      0         4           2         0       1    4               2   
4179      1         3           2         7       0    6               2   
319       1         3           3         8       0    1               2   
1380      0         2           3         8       0    7               2   
5345      1         2           0         0       1    6               2   
874       1         2           3         8       0    0               2   
8122      0         4           2         0       1  

Получим массивы с ответами

In [7]:
y_test = df_testing['edible'].copy().values.tolist()
y_learn = df_learning['edible'].copy().values.tolist()
print(len(y_test))
print(len(y_learn))
total = 0
for v in y_test:
    if v:
        total += 1
        
print(total/len(y_test))
total = 0

1160
6964
0.5086206896551724


Получим списки списков атрибутов

In [8]:
df_testing_copy = df_testing.copy()
df_learning_copy = df_learning.copy()
del df_testing_copy['edible']
del df_learning_copy['edible']
    
print(len(df_testing_copy.columns))    

x_test = df_testing_copy.values.tolist()
x_learn = df_learning_copy.values.tolist()

22


Проверим, что разделение прошло успешно

In [9]:
print((len(x_test) + len(x_learn) == len(df_numeric)))
print((len(x_test)))

True
1160


## Функционал

Датасет подготовлен для обучения модели. Введем несколько дополнительную функцию - эвклидово расстояние по которому будем рассчитывать расстояние

## Обучение модели

Обучать модель не нужно, knn достаточно предоставить базу данных с заранее известными ответами. Ответ будем получать сравнивая кол-во соседей заданной точки с заранее известными и затем выводить ответ в зависимости от того каких соседей больше.

In [10]:
from math import sqrt

class knn:
    def __init__(self, k=5, min_value=10):
        self.k = k
        self.min_value=min_value
    def euclidean(self, v1, v2):
        return sqrt(sum([(x - y)**2 for x, y in zip(v1, v2)]))
    
    def fit(self, X,Y):
        self.train_x = X
        self.train_y = Y
        return self
    
    def predict(self, X):
        edible_count = 0
        poison_count = 0
        result_list = list()
        while len(result_list) != self.k:
            min_value = self.min_value
            index = -1
            for i in range(len(self.train_x)):
                if i not in result_list:
                    ecld = self.euclidean(self.train_x[i], X)
                    if(min_value > ecld):
                        min_value = ecld
                        index = i
                        break
            if(index != -1):
                result_list.append(index)
            else:
                break
        
        for value in result_list:
            if (self.train_y[value] == 1):
                edible_count += 1
            else:
                poison_count += 1
        return 1 if edible_count > poison_count else 0
                        
                    
    
    def score(self, X, Y):
        success_count = 0
        for x, y in zip(X,Y):
            if self.predict(x) == y:
                success_count+=1
        return success_count/len(X)

### Тестирование

In [11]:
print("Точность алгоритма: {}%".format(knn(k=7, min_value=5).fit(x_learn, y_learn).score(x_test, y_test)*100))

Точность алгоритма: 98.1896551724138%


## Наивный байесовский классификатор

Реализуем простейший байесовский классификатор

Посчитаем вероятности встретить каждый признак в каждом классе

In [12]:
class bayes:
    def get_prob_map(self, train_x, train_y):
        result_map = {}
        classes_count = {}
        # посчитаем количество объектов всех возможных классов
        for i in range(len(train_y)):
            cur = train_y[i]
            if cur not in classes_count:
                classes_count[cur] = 1
            else:
                classes_count[cur] += 1
        # найдем вероятности всех атрибутов встретится в каждом классе
        for i in range(len(train_x)):
            row = train_x[i]
            ans = train_y[i]
            
            if ans not in result_map:
                result_map[ans] = []
                cur_list = result_map[ans]
                while len(cur_list) != len(row):
                    cur_list.append({})
            
            cur_list = result_map[ans]
            for j in range(len(row)):
                value = row[j]
                cur_dict = cur_list[j]
                if value not in cur_dict:
                    cur_dict[value] = 1
                else:
                    cur_dict[value] += 1
        for key in result_map:
            current_list = result_map[key]
            for i in range(len(current_list)):
                cur_dict = current_list[i]
                for value in cur_dict:
                    cur_dict[value] /= classes_count[key]
        
        return result_map
    
    def fit(self, X,Y):
        self.prob_map = self.get_prob_map(X,Y)
        return self
    
    def predict(self, X):
        result_dict = {}
        for i in range(len(X)):
            for key in self.prob_map:
                value = 0
                if X[i] in self.prob_map[key][i]:
                    value = self.prob_map[key][i][X[i]]
                if key in result_dict:
                    result_dict[key] += value
                else:
                    result_dict[key] = value
        max_value = 0
        answer = None
        for key in result_dict:
            if max_value < result_dict[key]:
                answer = key
                max_value = result_dict[key]
        return answer
    def score(self, X, Y):
        success = 0
        for x, y in zip(X,Y):
            if self.predict(x) == y:
                success += 1
        return success/len(X)

### Тестирование

In [13]:
print("Точность алгоритма: {}%".format(bayes().fit(x_learn, y_learn).score(x_test, y_test)*100))

Точность алгоритма: 87.5%


## Сравнение с реализацией sklearn

### Алгоритм KNN

Проводим тестирование

In [14]:
from sklearn.neighbors import KNeighborsClassifier
print("Точность алгоритма : {}%".format(KNeighborsClassifier().fit(x_learn, y_learn).score(x_test, y_test)*100))

Точность алгоритма : 100.0%


### Байесовский классификатор

Проводим тестирование

In [15]:
from sklearn.naive_bayes import GaussianNB

print("Точность алгоритма : {}%".format(GaussianNB().fit(x_learn, y_learn).score(x_test,y_test)*100))

Точность алгоритма : 88.44827586206897%


## Выводы

В ходе лабораторной работы были реализованы простейшие классификаторы KNN и байесовкий с приблизительной точностью для выбранного мной датасета в 98% и 87.5% соотвественно. Сравнил собственную реализацию с реализациями в библиотеке sklearn. Точность KNN собственной реализации оказалась довольно высокой, однако скорость вычисления класса - низкая, собственная реализация алгоритма байесовского классифакатора выдает практически такую же точность, что и в sklearn.

В ходе выполнения работы смог убедится на практике в том, что при помощи математики, вычислительной техники и данных можно находить различные корреляции между метриками, которые не поддаются обычному разъяснению для единичных сущностей, иначе говоря, на основе статистики можно с большой долей вероятности предсказать некоторые свойства выбранного объекта. 